source/docs/design/rw2022-talk-proposal.tex

```diff
@@ -0,0 +1,199 @@
+specifications and proofs.
+
+\section{Verus' design}\label{verus-design}
+
+Verus leverages the fact that safe Rust has a natural, direct encoding
+in SMT (for code without interior mutability, which we need to address
+separately). Algebraic data types (ADTs) and sound generics complement
+safe Rust to make it a solid basis for a mathematical language for
+expressing specifications and proofs. When functions represent theorems,
+expressing complex proofs like lemmas by induction is straightforward
+via a simple recursive call.
+
+Other tools, like Dafny, need complex encodings in SMT to support the
+semantics of the language (such as heap reasoning), and in our
+experience this can be the cause of unpredictable prover performance. We
+choose to prioritize efficient, straighforward encoding of Rust to Z3 in
+Verus (without an intermediate language like Boogie). This design
+principle comes with a necessary compromise: we do not intend to support
+all Rust features and libraries, instead excluding those that would
+require complicated encoding. We do however plan to provide
+verification-aided safe alternatives when unsafe is currently required
+in Rust.
+
+An explicit goal of the project is to build a solid, well-engineered
+foundation with excellent diagnostics, to aid onboarding of new users,
+in the spirit of the Rust's compiler error reporting (I will demonstrate
+this in the demos). A well-engineered foundation should also allow us
+and others to use Verus as the basis to experiment with new verification
+language features and techniques.
+
+\section{The type system and borrow checker}\label{the-type-system-and-borrow-checker}
+
+We consider the Rust's type system and borrow checker a central
+component in Verus' design and ability to generate verficiation
+conditions that Z3 can solve efficiently. Linear types mean that we can
+often encode potentially mutable, complex objects, as immutable SMT
+datatypes, rather than having to rely on complex rules and axioms. Even
+objects with interior mutability that expose an ``immutable'',
+deterministic interface can be often safely encoded as Z3 values. The
+borrow checker ensures that each value is either (i) currently uniquely
+owned, (ii) borrowed many times but immutably, or (iii) borrowed mutably
+exactly once.
+
+\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}
+
+Example. A simple Rust function with a functional spec and the generated
+Z3 verification condition.
+
+Rust:
+
+\begin{verbatim}
+fn small_sum(a: u64, b: u64) -> u64 {
+    requires([
+        a < 128,
+        b < 128,
+    ]);
+    ensures(|result: u64| result < 256);
+    a + b
+}
+\end{verbatim}
+
+Z3:
+
+\begin{verbatim}
+  (declare-const result@ Int)
+  (declare-const a@ Int)
+  (declare-const b@ Int)
+  ... (3 lines with heuristics machinery and u64 type invariants) ...
+  (assert (< a@ 128))
+  (assert (< b@ 128))
+  (assert
+   (not (=>
+     (= result@ (uClip 64 (+ a@ b@))) (< result@ 256)
+  )))
+\end{verbatim}
```

Comment:
it would be nice to visually show how Z3 lines correspond to the rust lines.

maybe color-code the lines so the reader can easily see how they correspond?

Reply:
I think we're going to remove the AIR for space reasons, but that's a good idea (for the talk).